src/components/ImagePreviewer.js

```diff
@@ -0,0 +1,176 @@
+import { stringify } from "querystring";
+import { decode } from "base58";
+
+const noop = () => {};
+
+export const of = src => Promise.resolve({ src });
+
+export const makeSrcResolver = ({ src }) =>
+  resolvers.reduceRight((acc, resolver) => {
+    if (!acc && resolver.test(src)) {
+      return resolver.request(src);
+    }
+    return acc;
+  }, false);
+
+export const makeImageDOMResolver = ({ src }) =>
+  new Promise((resolve, reject) => {
+    const img = new Image();
+    img.onload = () =>
+      resolve({
+        src,
+        height: img.height
+      });
+    img.onerror = reject;
+    img.src = src;
+  });
+
+export class ImagePreviewer extends React.PureComponent {
+  state = {
+    pending: undefined,
+    value: undefined
+  };
+
+  componentDidMount() {
+    this.handleStart();
+  }
+
+  componentDidUpdate(prevProps) {
+    if (this.props.request !== prevProps.request) {
+      this.handleStart();
+    }
+  }
+
+  handleStart(props) {
+    this.setState((state, { request }) => {
+      request.then(this.handleResolve, noop);
+      return {
+        pending: request,
+        value: undefined
+      };
+    });
+  }
+
+  handleResolve = value => {
+    this.setState(({ pending }, { request }) => {
+      if (pending !== request) {
+        return;
+      }
+      return { value };
+    });
+  };
+
+  render() {
+    return this.props.render({
+      ...this.props,
+      request: undefined,
+      ...this.state.value
+    });
+  }
+}
+
+const getTop = (top, height) => {
+  const pageHeight = $(window).height();
+
+  // opening image would pass the bottom of the page
+  if (top + height / 2 > pageHeight - 20) {
+    if (height / 2 < top) {
+      return pageHeight - 20 - height;
+    }
+  } else if (top - 20 > height / 2) {
+    return top - height / 2;
+  }
+  return 20;
+};
+
+ImagePreviewer.renderOnHover = ({ left, top, src, height }) => (
```

Comment:
Maybe move the two branch out. So this could be
`{src ? <Image {...props}> : <Spinner {...props}>}`

Reply:
Yeah but again, I don't meant to divide React component extensively at the moment. Just want to keep it as simple as possible and do the refactoring in the future.

Reply:
Ack.